### PR TITLE
SSCS-5054 - Subscription updated

### PR DESCRIPTION
**Pull request by ryandac**
The Subscription updated notification didn't work properly.

This PR is to get it working for appellant, representative and appointee.

* unsubscribe - does not send any notification
* subscribe email/mobile - will also resend the last notification
* change email/mobile

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/sscs/service/NotificationService.java [hunk 1]
```diff
@@ -60,51 +67,91 @@ private void sendNotificationPerSubscription(NotificationWrapper notificationWra
...
+                sendNotification(notificationWrapper, subscriptionWithType);
                 reminderService.createReminders(notificationWrapper);
+                resendLastNotification(notificationWrapper, subscriptionWithType);
             }
         }
     }
 
+    private void resendLastNotification(NotificationWrapper notificationWrapper, SubscriptionWithType subscriptionWithType) {
+        if (shouldProcessLastNotification(notificationWrapper, subscriptionWithType)) {
+            log.info("resending the last notification");
```
Comment (jackmaloney) [inline src/main/java/uk/gov/hmcts/reform/sscs/service/NotificationService.java]: Can we add some more details to this log message so we know what notification is resent and which case id?

File: src/main/java/uk/gov/hmcts/reform/sscs/service/NotificationService.java [hunk 2]
```diff
@@ -60,51 +67,91 @@ private void sendNotificationPerSubscription(NotificationWrapper notificationWra
...
+                || (!oldSubscription.isSmsSubscribed() && newSubscription.isSmsSubscribed()));
+    }
+
+    private static boolean thereIsALastEventThatIsNotSubscriptionUpdated(final SscsCaseData newSscsCaseData) {
+        boolean thereIsALastEventThatIsNotSubscriptionUpdated = newSscsCaseData.getEvents() != null
+                && !newSscsCaseData.getEvents().isEmpty()
+                && newSscsCaseData.getEvents().get(0).getValue().getEventType() != null
+                && !SUBSCRIPTION_UPDATED.equals(newSscsCaseData.getEvents().get(0).getValue().getEventType());
+        if (!thereIsALastEventThatIsNotSubscriptionUpdated) {
+            log.info("Not re-sending the last subscription as there is no last event on the case.");
```
Comment (jackmaloney) [inline src/main/java/uk/gov/hmcts/reform/sscs/service/NotificationService.java]: Same comment about the logs here...

File: src/main/java/uk/gov/hmcts/reform/sscs/service/NotificationService.java [hunk 3]
```diff
@@ -60,51 +67,91 @@ private void sendNotificationPerSubscription(NotificationWrapper notificationWra
...
+        Subscription oldSubscription = getSubscription(notificationWrapper.getOldSscsCaseData(), subscriptionWithType.getSubscriptionType());
+        Subscription newSubscription = subscriptionWithType.getSubscription();
+        String email = oldSubscription.isEmailSubscribed() ? null : newSubscription.getEmail();
+        String mobile = oldSubscription.isSmsSubscribed() ? null : newSubscription.getMobile();
+        subscriptionWithType.setSubscription(newSubscription.toBuilder().email(email).mobile(mobile).build());
+    }
+
+    private boolean shouldProcessLastNotification(NotificationWrapper notificationWrapper, SubscriptionWithType subscriptionWithType) {
+        return SUBSCRIPTION_UPDATED_NOTIFICATION.equals(notificationWrapper.getSscsCaseDataWrapper().getNotificationEventType())
+                && !isPaperCase(notificationWrapper.getNewSscsCaseData().getAppeal().getHearingType())
```
Comment (jackmaloney) [inline src/main/java/uk/gov/hmcts/reform/sscs/service/NotificationService.java]: Where is the requirement to not send the last notification if it's a paper case? Can't see that on the ticket so just want to make sure this is intended...

File: src/test/java/uk/gov/hmcts/reform/sscs/factory/NotificationFactoryTest.java [hunk 4]
```diff
@@ -248,41 +253,47 @@ public void buildLastNotificationFromSscsCaseDataEventWhenSmsFirstSubscribed() {
...
+                                .subscriptions(Subscriptions.builder().appellantSubscription(subscription.toBuilder().subscribeSms("Yes").subscribeEmail("Yes").build()).build())
                                 .build())
                 .notificationEventType(SUBSCRIPTION_UPDATED_NOTIFICATION)
                 .build();
 
-        Notification result = factory.create(new CcdNotificationWrapper(wrapper), APPELLANT);
+        Notification result = factory.create(new CcdNotificationWrapper(wrapper), getSubscriptionWithType(wrapper, APPELLANT));
 
-        assertNull(result.getEmailTemplate());
+        assertEquals("123", result.getEmailTemplate());
```
Comment (jackmaloney) [inline src/test/java/uk/gov/hmcts/reform/sscs/factory/NotificationFactoryTest.java]: This doesn't look correct to me? The test name doesn't want a notification to be sent but you have changed the assertions to check that an email and sms template exist.

File: src/IntegrationTests/java/uk/gov/hmcts/reform/sscs/tya/NotificationsIt.java [hunk 5]
```diff
@@ -903,7 +898,7 @@ public void shouldSendSubscriptionUpdatedNotificationForSubscriptionUpdatedReque
         HttpServletResponse response = getResponse(getRequestWithAuthHeader(json));
 
         assertHttpStatus(response, HttpStatus.OK);
-        verify(notificationClient, times(1)).sendEmail(any(), any(), any(), any());
+        verify(notificationClient, times(2)).sendEmail(any(), any(), any(), any());
```
Comment (jackmaloney) [inline src/IntegrationTests/java/uk/gov/hmcts/reform/sscs/tya/NotificationsIt.java]: Is there any way we can break this down to verify what types of emails are sent here so we can be sure it's sending the correct ones?

File: src/test/java/uk/gov/hmcts/reform/sscs/factory/NotificationFactoryTest.java (hunk 4 above)
Comment (ryandac) [reply src/test/java/uk/gov/hmcts/reform/sscs/factory/NotificationFactoryTest.java]: It used to change the notification event to `donotsend` and hence has a null email and sms template.
Now, It doesn't get sent due to the email and sms being null 2 lines below.

File: src/main/java/uk/gov/hmcts/reform/sscs/service/NotificationService.java (hunk 3 above)
Comment (ryandac) [reply src/main/java/uk/gov/hmcts/reform/sscs/service/NotificationService.java]: I copied it from existing code. The existing code had a lot of bugs. Looking at the acceptance criteria https://tools.hmcts.net/jira/browse/SSCS-3955. This doesn't have the paper case requirement so I'll remove it.

**Pull request merged**